### ISSUE TITLE
Translation Plan

### ISSUE THREAD

**Issue by meowmeowmeowcat** [pypa/packaging.python.org#927]
Hi,

I would like to ask if there is any translation plan of packaging.python.org?

**Comment by webknjaz** [comment]
Hi @meowmeowmeowcat, no. At the moment nobody raised this question AFAIK. But I like the idea and I think it would be great to set up a translation process. This may require a discussion with other maintainers first, but I'd definitely support such an effort.

**Comment by webknjaz** [comment]
@pypa/packaging-user-guide-editors @ewdurbin @di @pradyunsg do you think it'd be possible to use Warehouse's Weblate for this? I think the configuration pointers are present here https://docs.weblate.org/en/latest/devel/sphinx.html.

**Comment by webknjaz** [comment]
@meowmeowmeowcat as a side note, the translation would require a dedicated person per language to continuously update the articles as the canonical English versions change and new articles appear. Would you commit to something like this? You'd need to take ownership over curating your language.

**Comment by pradyunsg** [comment]
> I would like to ask if there is any translation plan of packaging.python.org?

None at this time, no.

Sphinx and ReadTheDocs have fairly strong internationalization support -- so if someone is interested in translating and maintaining that translated guide, I'm pretty sure we can make that happen. Personally, I wouldn't be pushing for setting up that infrastructure and figuring those details out, unless someone is actually interested in doing the translation.

**Comment by anxolerd** [comment]
It would be nice to have packaging.python.org in different languages. I would even contribute to Ukrainian translation time to time, but, unfortunately i cannot constantly look at the original English articles and update translation on daily or weekly basis

**Comment by meowmeowmeowcat** [comment]
> @meowmeowmeowcat as a side note, the translation would require a dedicated person per language to continuously update the articles as the canonical English versions change and new articles appear. Would you commit to something like this? You'd need to take ownership over curating your language.

Sure, and I have translated some English to `zh_Hant` and `zh_Hans` on [PyPI](http://pypi.org) actually.

**Comment by meowmeowmeowcat** [comment]
> None at this time, no.
> 
> Sphinx and ReadTheDocs have fairly strong internationalization support -- so if someone is interested in translating and maintaining that translated guide, I'm pretty sure we can make that happen. Personally, I wouldn't be pushing for setting up that infrastructure and figuring those details out, unless someone is actually interested in doing the translation.

Actually I am interested in doing the translation and that's why I open this issue. 🤣

**Comment by webknjaz** [comment]
@meowmeowmeowcat it would be useful if somebody checked what we need to start doing this and maybe outlined the mechanics of the whole process. PRs adding the necessary configuration (including CI checks and a guide for translators into existing and new languages) are welcome too.

**Comment by meowmeowmeowcat** [comment]
@webknjaz Is it possible to create a new translation project under [pypa on weblate](https://hosted.weblate.org/projects/pypa/)?

**Comment by meowmeowmeowcat** [comment]
I have already made a example on [Weblate](https://hosted.weblate.org/projects/packaging-python-org/packaging-python-org/)

**Comment by meowmeowmeowcat** [comment]
So I think here are the steps we need to start doing the translation:

- We should generate `.po` of the docs for translation (which is done)
- Put `.po` files on the Weblate
- Add Translation CI check
- Writing a guide for Translators

**Comment by webknjaz** [comment]
> @webknjaz Is it possible to create a new translation project under [pypa on weblate](https://hosted.weblate.org/projects/pypa/)?

Probably yes. But I don't have such access. That's why I've tagged @di and @ewdurbin earlier — they should have a better answer to how this could be implemented.

**Comment by webknjaz** [comment]
> * We should generate `.po` of the docs for translation (which is done)
> * Put `.po` files on the Weblate

I _think_ the Weblate bot (integration?) is able to generate those files if set up properly. My understanding is that we'll need to generate them often or even embed this into the merging process. I hope folks who have set it up for Warehouse could shed some light on how it's done.

**Comment by meowmeowmeowcat** [comment]
> > @webknjaz Is it possible to create a new translation project under [pypa on weblate](https://hosted.weblate.org/projects/pypa/)?
> 
> Probably yes. But I don't have such access. That's why I've tagged @di and @ewdurbin earlier — they should have a better answer to how this could be implemented.

Okay. Thanks

**Comment by meowmeowmeowcat** [comment]
> I _think_ the Weblate bot (integration?) is able to generate those files if set up properly. My understanding is that we'll need to generate them often or even embed this into the merging process. I hope folks who have set it up for Warehouse could shed some light on how it's done.

So probably we should wait someone answering this before we go to the next step

**Comment by meowmeowmeowcat** [comment]
> > @pypa/packaging-user-guide-editors @ewdurbin @di @pradyunsg do you think it'd be possible to use Warehouse's Weblate for this? I think the configuration pointers are present here https://docs.weblate.org/en/latest/devel/sphinx.html.
> 
> I think we could add another component under the pypa project on weblate, yes. I'll contact them to ensure that would be OK.

Thanks!

**Comment by meowmeowmeowcat** [comment]
@ewdurbin Could you add another component under the PyPA project on Weblate, please? The template of translation is located in `locales/messages.pot`.

**Comment by pradyunsg** [comment]
> The template of translation is located in `locales/messages.pot`.

Let's move the generation of this into a GitHub action and stop checking this file into the version control -- the output of this is purely deterministic, and checking it in doesn't really have any direct benefit AFAICT.

**Comment by meowmeowmeowcat** [comment]
I think so.

**Comment by webknjaz** [comment]
@pradyunsg I'm pretty sure it needs to be checked in because Weblate is supposed to pick up the phrases from it per my understanding, I'm pretty sure Warehouse works like this.

**Comment by ewdurbin** [comment]
https://hosted.weblate.org/projects/pypa/packaging-python-org/ has been created. I'd like to ask for one or two folks involved in managing the translation to create accounts on weblate and share their usernames here so I can add them to administer the component.

**Comment by pradyunsg** [comment]
> I'm pretty sure it needs to be checked in

Okay, that sucks. For starters, we'll still need the same automation -- to ensure that it's properly generated on each revision. 

It's also a pretty bad state of "updates will need more than just changing the text" in our workflows IMO since it eliminates easy GitHub-UI only changes as well. As an easy example, we'd no longer be able to take GitHub UI driven typo fixes, which... is pretty bad IMO. And, this overhead would likely result in me being significantly less likely to actually write/update the content here myself -- admittedly I haven't done much so we're not losing out on much here, but that might change if other more-active contributors feel similarly.

**Comment by webknjaz** [comment]
@pradyunsg

> Okay, that sucks. For starters, it doesn't change that we'd need the same automation -- but it'll instead ensure that it's properly generated on each revision instead of just generating it.

I think we could come up with some solution pushing the updates to GH automatically, on PR merge, for example. Also, not having the extracted strings in main up-to-date shouldn't be a big deal, we could make sure to update it periodically for starters.


@ewdurbin 

> I'd like to ask for one or two folks involved in managing the translation to create accounts on weblate and share their usernames here so I can add them to administer the component.

@anxolerd and I have the same usernames as on GitHub, we both worked on the Ukrainian translation of Warehouse. I guess we'd need somebody from a different language background added later.

**Comment by meowmeowmeowcat** [comment]
@webknjaz @pradyunsg Is it possible for us to update `messages.pot` automatically by Github Action when merging the PR?

**Comment by webknjaz** [comment]
That's what I was thinking

**Comment by meowmeowmeowcat** [comment]
@webknjaz I've already prepared the Github Action file. However, you may need to generate a secret called `ACCESS_TOKEN` in `Settings/Secrets` for Github Action to push changes of `messages.pot`

**Comment by meowmeowmeowcat** [comment]
@webknjaz I think a better solution is open a new repo (probably called `packaging.python.org-l10n`) for translations, so the new repo will receive changes in packaging.python.org and update `messages.pot` constantly. Just like python docs.

**Comment by webknjaz** [comment]
I considered it but I don't see many problems with having a single repo. If it's autoupdated periodically, separately from PRs, it shouldn't cause any disturbance.

**Comment by meowmeowmeowcat** [comment]
So there would be a lot commits from Github Action (probably I should set it to update automatically every three/four days so there won't have so many commits), is this acceptable for this repo? :)

**Comment by ewdurbin** [comment]
`webknjaz` and `anxolerd` users on Weblate have been added and should have administrative control over the whole PyPA project on Weblate. @webknjaz and @anxolerd please do not modify settings for the warehouse component or glossary without consulting the pypa/warehouse issue tracker.

I'm going to unsubscribe from this thread now. Please mention me again if there is something I need to do.

**Comment by webknjaz** [comment]
> So there would be a lot commits from Github Action (probably I should set it to update automatically every three/four days so there won't have so many commits), is this acceptable for this repo? :)

Absolutely. Note that using a pre-populated `GITHUB_TOKEN` secret will not generate any events on GitHub meaning there will be no test runs triggered and no GitHub Apps will react to those commits.

**Comment by meowmeowmeowcat** [comment]
Thanks for answering this!

**Comment by meowmeowmeowcat** [comment]
@webknjaz Could you please change the language code on Weblate to be identical with readthedocs?

**Comment by webknjaz** [comment]
@meowmeowmeowcat we need to change the automation to produce paths without `../` in front of the collected strings: this breaks links from Weblate to GH currently (because it removes the segment with the branch name from URLs).

**Comment by meowmeowmeowcat** [comment]
@webknjaz Can this be fixed in Weblate settings?

**Comment by rffontenelle** [comment]
Hey there. Thanks for the initiative of internationalizing packaging.python.org guide. I'd like to make a suggestion:

While looking for instructions on building translated docs locally, I notice that [p-p-o info page on Weblate](https://hosted.weblate.org/projects/pypa/packaging-python-org/pt_BR/#information) seems to have exact [info as Warehouse component](https://hosted.weblate.org/projects/pypa/warehouse/pt_BR/#information). Not sure if that is a project-wide info, but would be nice to adapt p-p-o info page to be more useful for those localizing p-p-o.

Also, the info has RST syntax (should I take this to Warehouse?) but only Markdown is accepted, so I recommend writing p-p-o's instructions in Markdown.

**Comment by webknjaz** [comment]
@rffontenelle so I looked into it and there's no per-component setting, it's project-global. And is edited at https://hosted.weblate.org/settings/pypa/ — @ewdurbin @di what can we do about this? Should we mention both projects there?

**Comment by webknjaz** [comment]
@meowmeowmeowcat if you inspect the URLs in `<a href="...">` with DevTools, you'll see that the branch name **is included**.
One example URL is `https://github.com/pypa/packaging.python.org/blob/main/../source/overview.rst#L61`. But because of `/../`, it gets normalized as `https://github.com/pypa/packaging.python.org/blob/source/overview.rst#L61` which is broken.
The only proper way to fix this is to update the pot file not to contain `../` in front of each reference.

**Comment by meowmeowmeowcat** [comment]
@webknjaz I've checked some other docs and they didn't remove the `../` in front of each reference, but they still work properly. Therefore, I think this could be fixed in settings. 

> The only proper way to fix this is to update the pot file not to contain ../ in front of each reference.

Also, I can't remove`../` because the `messages.pot` is located in `locales/`, which means that it must use `../` to get the text in the docs.

**Comment by meowmeowmeowcat** [comment]
Could you please change the original one in Repository Browser to `https://github.com/pypa/blob/{{branch}}/{{filename}}#L{{line}}`?

**Comment by webknjaz** [comment]
I guess I could add that `/source/` chunk but it just looks like a weird workaround rather than a proper solution.

**Comment by webknjaz** [comment]
btw Warehouse doesn't have `../`: you can see it in https://github.com/pypa/warehouse/blob/main/warehouse/locale/messages.pot.

**Comment by meowmeowmeowcat** [comment]
> I guess I could add that /source/ chunk but it just looks like a weird workaround rather than a proper solution.

I agree. However, let's use this solution first until we find a better solution for this.

**Comment by meowmeowmeowcat** [comment]
@webknjaz I've found a solution:

By putting the `messages.pot` in dir `./`, and the `.pot` won't have `../` anymore.

Example:

```
#: source/guides/distributing-packages-using-setuptools.rst:95
msgid "For an example, see `README.md <https://github.com/pypa/sampleproject/blob/master/README.md>`_ from the `PyPA sample project <https://github.com/pypa/sampleproject>`_."
msgstr ""
```

**Comment by webknjaz** [comment]
I currently injected `/locales/`, will address this later then.

**Comment by meowmeowmeowcat** [comment]
@webknjaz Could you explain more for injecting `/locales/`?

**Comment by webknjaz** [comment]
@meowmeowmeowcat it's the same as your suggestion to add `/source/` but is more correct technically because the path is relative to the pot file.

**Comment by meowmeowmeowcat** [comment]
Oh, I see.  :)

**Comment by webknjaz** [comment]
@rffontenelle @meowmeowmeowcat looking at your translation attempts I must urge you not to translate references. I saw some translations changing the text in places like
```
`some ref`_
```
But this sort of ref is not just a rendered text, it's also bound to the link somewhere in the document that looks like
```rst
.. _some ref:: https://...
```
And changing the ref in one place will break this connection making sphinx incapable of finding the reference target.

Instead, I suggest using an explicit syntax for separate anchor text and the reference identifier that would look as follows:
```
`TRANSLATED TEXT HERE <some ref>`_
```

**Comment by webknjaz** [comment]
On a side note, today I've enabled syntax highlighting in the translation fields which should make it easier to make sure that it doesn't have broken sequences.

**Comment by meowmeowmeowcat** [comment]
@webknjaz Thanks for your reminder! I think we should add this to our translation guide as folks may make mistake like this.

**Comment by webknjaz** [comment]
Yes, that's an excellent idea. I just wanted to make sure that somebody fixes the mistakes that have been made already.

**Comment by meowmeowmeowcat** [comment]
>  I just wanted to make sure that somebody fixes the mistakes that have been made already.

I've fixed mistakes that I have been made. I also marked some translations in `pt_BR` as need edit.

@rffontenelle Please fix them.

**Comment by ewdurbin** [comment]
> @rffontenelle so I looked into it and there's no per-component setting, it's project-global. And is edited at https://hosted.weblate.org/settings/pypa/ — @ewdurbin @di what can we do about this? Should we mention both projects there?

That's copy/pasted from warehouse's docs. We could probably just update it to link to https://warehouse.readthedocs.io/translations.html for warehouse and some page on packaging.python.org if you want to create it once a workflow is settled on.

**Comment by rffontenelle** [comment]
@webknjaz @meowmeowmeowcat thanks for the heads up. Fixed the ref links.

One nice syntax guide for RST docs: https://docutils.sourceforge.io/docs/user/rst/quickref.html

**Comment by meowmeowmeowcat** [comment]
> One nice syntax guide for RST docs: https://docutils.sourceforge.io/docs/user/rst/quickref.html

@rffontenelle Thanks for suggestion. This guide will be added to translation guide.

**Comment by webknjaz** [comment]
@ewdurbin @di there's an alert here https://hosted.weblate.org/projects/pypa/packaging-python-org/#alerts and the :warning: symbol is displayed next to our project. It seems to imply that the PSF license is unacceptable. Could you take a look?

**Comment by rffontenelle** [comment]
@webknjaz [It seems] that Weblate relies on [SPDX License List] to tell if the license is accepted or not, which means being FSF Free/Libre and OSI Approved. This is done [automagically] by extracting for SPDX License List, and that list has "Python License 2.0" with Y(es) for both requirements, and "Python Software Foundation License 2.0" which does **not** have Y for both requirements. PyPA projects on Weblate are set with "Python Software Foundation License 2.0" as license.

If this is a incorrect result, consider filing a bug report in https://github.com/WeblateOrg/weblate/ to work this out.

[It seems]: https://github.com/WeblateOrg/weblate/issues/4932#issuecomment-735222663
[SPDX License List]: https://spdx.org/licenses/
[automagically]: https://github.com/WeblateOrg/weblate/blob/main/scripts/generate-license-data

**Comment by ewdurbin** [comment]
looks like @rffontenelle's analysis is correct: https://opensource.org/licenses/Python-2.0 https://www.gnu.org/licenses/license-list.en.html#Python indicate that the license is FSF and OSI approved. Might be something we need to discuss with SPDX too?

**Comment by webknjaz** [comment]
> Might be something we need to discuss with SPDX too?

Yep, that's what I was thinking too.

**Comment by rffontenelle** [comment]
> looks like @rffontenelle's analysis is correct: https://opensource.org/licenses/Python-2.0 https://www.gnu.org/licenses/license-list.en.html#Python indicate that the license is FSF and OSI approved. Might be something we need to discuss with SPDX too?

Let's differentiate "Python-2.0" from "Python Software Foundation License 2.0" in SPDX's point of view.

You are saying "Python-2.0", but "Python Software Foundation License 2.0" is set in Weblate, which is "PSF-2.0" according to SPDX. To SPDX, the following are two different licenses:

**Full name: Python Software Foundation License 2.0**
Identifier: PSF-2.0
SDPX URL: https://spdx.org/licenses/PSF-2.0.html
FSF Free/Libre: (empty)
OSI Approved: (empty)

**Full name: Python License 2.0**
Identifier: Python-2.0
SDPX URL: https://spdx.org/licenses/Python-2.0.html
FSF Free/Libre: Y
OSI Approved: Y

It seems you are expecting "Python-2.0" to be accepted (which is!), but the projects in Weblate were actually set with PSF-2.0's full name.

**Comment by rffontenelle** [comment]
Also, what's with CC BY-SA 3.0 license in this project's README file ?
> The Python Packaging User Guide is licensed under a Creative Commons Attribution-ShareAlike license: http://creativecommons.org/licenses/by-sa/3.0.

https://packaging.python.org/'s footnote states "This page is licensed under the Python Software Foundation License Version 2."

**Comment by webknjaz** [comment]
> You are saying "Python-2.0", but "Python Software Foundation License 2.0" is set in Weblate, which is "PSF-2.0" according to SPDX.

It's interesting that both SPDX pages point at https://opensource.org/licenses/Python-2.0 in the `Other web pages for this license` field. And there's also this note:
> This is the PSF-2.0 license, which is part of the complete Python license text, but also used independently by some projects.

So basically if `Python-2.0` is libre, then its part (PSF-2.0) should also be libre...

**Comment by webknjaz** [comment]
> [packaging.python.org](https://packaging.python.org/) 's footnote states "This page is licensed under the Python Software Foundation License Version 2."

Looks like this is coming from the theme and is not overridden in this repo: https://github.com/python/python-docs-theme/blob/6ed468a/python_docs_theme/layout.html#L122-L125.
I wonder if @pradyunsg's Furo will have the same problem — adopting it (or its derivative) will probably just drop the license mention completely).

**Comment by meowmeowmeowcat** [comment]
> On a side note, today I've enabled syntax highlighting in the translation fields which should make it easier to make sure that it doesn't have broken sequences.

@webknjaz Is syntax highlighting disabled?

**Comment by webknjaz** [comment]
It should be enabled.

**Comment by rffontenelle** [comment]
I see it enabled in pt_BR

**Comment by meowmeowmeowcat** [comment]
<img width="1209" alt="image" src="https://user-images.githubusercontent.com/68463158/132088368-e753b3a7-7279-435b-bcc7-36b4d8d447ba.png">

<img width="1296" alt="image" src="https://user-images.githubusercontent.com/68463158/132088404-c9180b33-7118-4f07-8900-0f0f69865112.png">

It sounds very strange now. I can see that Ukrainian has syntax highlighting, but other languages don't have it.

**Comment by rffontenelle** [comment]
@meowmeowmeowcat Can you please provide an example and link to a string of any language which doesn't have syntax highlighting?

**Comment by meowmeowmeowcat** [comment]
Sure. 

https://hosted.weblate.org/translate/pypa/packaging-python-org/zh_Hans/?q=state%3Aempty&sort_by=-priority%2Cposition&offset=700

https://hosted.weblate.org/translate/pypa/packaging-python-org/eo/?q=state%3Aempty&offset=5

**Comment by meowmeowmeowcat** [comment]
🤦 I know what's going on. There is no issue with syntax highlight. Sorry for the confusion.

**Comment by webknjaz** [comment]
FYI @pradyunsg and I tried to come up with a workaround for auto-updating the POT file (on a social call that we have periodically) and decided that the updates should happen in the PRs with changes to the respective RST files (for many reasons, including the fact that it's seemingly impossible to make GHA / bots bypass the branch projection). To facilitate this, we'll try integrating https://pre-commit.ci so that the contributors wouldn't need to do this manually (the bot will just push the updates to their branches effectively updating the PRs).

**Comment by meowmeowmeowcat** [comment]
Good idea! 

> To facilitate this, we'll try integrating https://pre-commit.ci so that the contributors wouldn't need to do this manually (the bot will just push the updates to their branches effectively updating the PRs).

@webknjaz Would you take care of this?

**Comment by webknjaz** [comment]
Yep, also Pradyun needs to add the integration, I don't have sufficient privileges for this.

**Comment by meowmeowmeowcat** [comment]
That's great! Thank you for both of you working on this!

**Comment by webknjaz** [comment]
@meowmeowmeowcat do you know if the MO files are necessary for the localization to work? I'd prefer not to keep binaries in the repo.

**Comment by webknjaz** [comment]
Ah, there's `gettext_auto_build`: https://github.com/readthedocs/readthedocs.org/issues/4702#issuecomment-426556170.

**Comment by meowmeowmeowcat** [comment]
> Ah, there's `gettext_auto_build`: https://github.com/readthedocs/readthedocs.org/issues/4702#issuecomment-426556170.

Yes, and you have removed `.mo` files from the repo. So this is solved. However, I still wonder if they are required to build the docs. I'll have a test later.

**Comment by webknjaz** [comment]
> > Ah, there's `gettext_auto_build`: [readthedocs/readthedocs.org#4702 (comment)](https://github.com/readthedocs/readthedocs.org/issues/4702#issuecomment-426556170).
> 
> Yes, and you have removed `.mo` files from the repo.

They are required but Sphinx generates them during the build. So it works (I've checked both locally and on RTD + added the respective setting explicitly).

**Comment by meowmeowmeowcat** [comment]
> @meowmeowmeowcat do you know if the MO files are necessary for the localization to work? I'd prefer not to keep binaries in the repo.

Great! So this is solved. :) Now I have another question: Can you build other languages for the docs, for example `zh_Hans` etc.

**Comment by webknjaz** [comment]
It looks like Sphinx doesn't validate the exact language codes. According to my tests, it accepts any string there.

**Comment by meowmeowmeowcat** [comment]
Oh, I see. And I've also found a solution for RTD. RTD follows the language codes given by Sphinx: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-language. After I checked the docs of Weblate, I think we don't have to modify our language codes now. We can add symlinks instead.

**Comment by rffontenelle** [comment]
I'm getting a check failure on a regex and I believe it shouldn't be failing. I assume the regex used is customized. Can this be verified? [This is the translation string](https://hosted.weblate.org/translate/pypa/packaging-python-org/pt_BR/?q=has:check) .

 (Click in the images to see full size)

<img src="https://user-images.githubusercontent.com/1571783/141664429-87f4f5b0-aa24-4ad4-889d-447889815876.jpg" alt="quality check failure on regex" width="70%" height="70%">

<img src="https://user-images.githubusercontent.com/1571783/141664437-d5211e5d-bfc0-4b3a-a349-f8187cf4dee0.jpg" alt="quality check failure on regex" width="70%" height="70%">

**Comment by meowmeowmeowcat** [comment]
@rffontenelle I guess translating the string
```rst
:ref:`python_requires`
```
to:
```rst
:ref:`translated_text <python_requires>`
```
should silence this error, as Weblate assumes that it would be translated.

**Comment by rffontenelle** [comment]
@meowmeowmeowcat you guessed correctly. However ``` :ref:`python_requires` ``` should be considered valid too to reStructuredText, and is normally not translated AFAIK.

How about adjusting the regex to something that matches both the current state and also the original text? E.g.

```
:ref:`(\w+[\w\s]* <python_requires>|python_requires)`
```

**Comment by meowmeowmeowcat** [comment]
@webknjaz Could you help us on modifying this if possible? Thanks!

**Comment by webknjaz** [comment]
@rffontenelle @meowmeowmeowcat thanks for bringing this up! Adding those checks was a lot of manual work, I haven't done this for all the strings (never got time) and I've tried to make them generic where possible. The common case is that you'd actually need a custom translation in many cases, for example, because if the linked label has a title with a different text, that title will be injected as-is. This is problematic for languages that would require that text be adapted, to be consistent with the sentence it's linked in. In the translation UI, there's no way to see what's on the "other side" of the label so I've opted for always requiring explicit text.

In your case, you're lucky that https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires got the same text as the label but have you checked this manually, or has it happened by chance? The principle that explicit is better than implicit applies here too, I think.

FWIW I've updated the check for that string.

**Comment by rffontenelle** [comment]
@webknjaz I thought it was a custom setting, but haven't thought on the work it took to set it up. Thanks for the effort, it is very appreciated.

It wasn't luck---I knew that python_requires was a term that shouldn't be translated. I normally have a tab with the translate strings in Weblate and another tab with the documentation. So I see what's the resulting the RST syntax.

But please notice that is it not mandatory to use explicit translation in reference to get translated result. When ref'ing a [hyperlink target](https://docutils.sourceforge.io/docs/user/rst/quickref.html#hyperlink-targets), the translated title is injected when such title was translated.

For instance, Weblated documentation has a `component` hyperlink target the gets replaced with "Component configuration" in original doc. However in pt_BR translation, the `component` target is replaced with "Configuração de componente" **without expliciting such translated title**. E.g. "Veja :ref:\`component` para mais informações" would result in "Veja Configuração de componente para mais informações

Personally I prefer not translating the :ref: without obvious needs (e.g. plural/singular differences) to try to avoid inconsistencies (e.g. when changing the title translation).

**Comment by rffontenelle** [comment]
How can I build translated version of packaging.python.org?

While I successful build pt_BR version (confirmed due to Sphinx own translated messages), my translation in this project doesn't land in the HTML pages. I'm using the following command:

```
sphinx-build -b html -d build/doctrees --keep-going -jauto -D locale_dirs=locales -D language=pt_BR source build/html
```

**Comment by webknjaz** [comment]
> @webknjaz I thought it was a custom setting, but haven't thought on the work it took to set it up. Thanks for the effort, it is very appreciated.

Well, it is sorta custom depending on the PoV. It's implemented via "flags" that are documented here: https://docs.weblate.org/en/latest/admin/checks.html#custom-checks.
I've set some flags globally for the project, like the list of common terms/abbreviations. It gets inherited by all the strings but can be overridden for each string separately. I've gone through a bunch of those that have RST and added this regex that you saw.
Usually, you should see the flags in the sidebar. Here's an example:
![weblate-flags](https://user-images.githubusercontent.com/578543/141684321-4beb61da-53f6-4f81-bf4a-5ca2f1412e22.png)
And editing that looks like this:
![weblate-inheritance](https://user-images.githubusercontent.com/578543/141684340-158e2ed3-f691-440c-b11d-9c87c34f050b.png)

I'm trying to add these checks so that they are useful across languages and help many people.

> It wasn't luck---I knew that python_requires was a term that shouldn't be translated. I normally have a tab with the translate strings in Weblate and another tab with the documentation. So I see what's the resulting the RST syntax.
> 
> But please notice that is it not mandatory to use explicit translation in reference to get translated result. When ref'ing a [hyperlink target](https://docutils.sourceforge.io/docs/user/rst/quickref.html#hyperlink-targets), the translated title is injected when such title was translated.

This is exactly the problem. You don't see what's on the other side. Moreover, if the title under that ref changes in the future, this will happen separately from updating the references in many cases.

Currently, it's
```rst
.. _python_requires:

``python_requires``
~~~~~~~~~~~~~~~~~~~
```

meaning that it'll be `python_requires` in the rendered text:
```rst
use the :ref:`python_requires` argument
```
will become
```html
use the <a href=...>python_requires</a> argument
```
when rendered. But imagine it changes like this:
```diff
 .. _python_requires:

-``python_requires``
+``python_requires`` explanation
 ~~~~~~~~~~~~~~~~~~~
```
then the render will silently become
```html
use the <a href=...>python_requires explanation</a> argument
```
breaking the sentence structure/grammar/etc.

The translators will not get notified until somebody sees this on the published site which is what I'm trying to avoid by making sure that the sentences with the references are always consistent.

> For instance, Weblated documentation has a `component` hyperlink target the gets replaced with "Component configuration" in original doc. However in pt_BR translation, the `component` target is replaced with "Configuração de componente" **without expliciting such translated title**. E.g. "Veja :ref:`component` para mais informações" would result in "Veja Configuração de componente para mais informações
> 
> Personally I prefer not translating the :ref: without obvious needs (e.g. plural/singular differences) to try to avoid inconsistencies (e.g. when changing the title translation).

I know what you're talking about but many people would not notice the lack or the presence of this "obvious" need when going through the list of the strings during the translation and this burden then becomes translation editor's.

With automatic validation, the chance of getting into this sort of trouble is lower which is why I've gone for it in the first place.

**Comment by webknjaz** [comment]
> How can I build translated version of packaging.python.org?

I think I've added a `nox` target for this earlier. For Ukrainian, I used to run `python -m nox -s build -- -D language=uk` (just found it in my terminal history). Also, I've set up a separate RTD site with manual builds and a different language selected: https://packagingpythonorg-uk-ua.rtfd.io — it took a few tries to figure out the proper locale name matching what RTD has in their selector. Could do the same for pt-BR.

**Comment by webknjaz** [comment]
@rffontenelle I've created an RTD site for pt-BR here https://packagingpythonorg-pt-br.rtfd.io. The build seems to complete successfully: https://readthedocs.org/projects/packagingpythonorg-pt-br/builds/15270285/. If you log in to RTD and give me your username there, I'll add you to the project so you could experiment with rebuilding the site occasionally. This is only for testing, of course. When at least one of the languages gets 90-95% of the content translated, we'll think about how to best deploy this for wider consumption.

**Comment by rffontenelle** [comment]
@webknjaz

Regarding our ref talk above, you understood my idea. And I understand that if the target string changes ("python_requires explanation" in your example), will changes how it is rendered. My thinking on this is: I translate making sure it renders properly in the documentation, also with proper grammar. If the "python_requires" title changes, I'll translate it as well. It might require some change around the documentation if any. On the other side, imagine having to change all occurrences of "python_requires" in translation due to title change -- IMHO, it would require more work and very prone to inconsistency.

Let me promise this, now and then I will build the documentation with the up-to-date PO file (or using RTD) to make sure it doesn't have Sphinx WARNINGS -- I do this for pt_BR translation of Python Docs.

----

Regarding flags in Weblate UI, no, can't see the regex on my side. It is probably visible only to the project's admins.

----

My RTD account is rafaelff

Thanks for setting up a testing website. It will be very useful.

**Comment by rffontenelle** [comment]
For the record, that's how I locally built pt_BR version of packaging.python.org:

```bash
pip install nox    # if not installed already
nox -s translation
wget https://hosted.weblate.org/download/pypa/packaging-python-org/pt_BR/ -O locales/pt_BR/LC_MESSAGES/messages.po
msgmerge --update --previous locales/pt_BR/LC_MESSAGES/messages.po locales/messages.pot
nox -s build -- -D language=pt_BR
```

It should work for any other language, replacing pt_BR with the proper language code.

**Comment by webknjaz** [comment]
> Regarding our ref talk above, you understood my idea. And I understand that if the target string changes ("python_requires explanation" in your example), will changes how it is rendered. My thinking on this is: I translate making sure it renders properly in the documentation, also with proper grammar. If the "python_requires" title changes, I'll translate it as well. It might require some change around the documentation if any. On the other side, imagine having to change all occurrences of "python_requires" in translation due to title change -- IMHO, it would require more work and very prone to inconsistency.
> 
> Let me promise this, now and then I will build the documentation with the up-to-date PO file (or using RTD) to make sure it doesn't have Sphinx WARNINGS -- I do this for pt_BR translation of Python Docs.

I see. Well, my thinking was that this approach would let people not care about the title updates since the context of where the references are used probably wouldn't change and wouldn't require changed wording. I'd argue that unanchored refs shouldn't be used in the source either because of the untraceable implications of the side effects caused by this.

> Regarding flags in Weblate UI, no, can't see the regex on my side. It is probably visible only to the project's admins.

Ah, I wasn't sure about that, this is why I've added those screenshots. But if you see something that may benefit from extra validation, let me know.

> My RTD account is rafaelff
> 
> Thanks for setting up a testing website. It will be very useful.

Invited.

**Comment by webknjaz** [comment]
> For the record, that's how I locally built pt_BR version of packaging.python.org:
> 
> ```shell
> pip install nox    # if not installed already
> nox -s translation
> wget https://hosted.weblate.org/download/pypa/packaging-python-org/pt_BR/ -O locales/pt_BR/LC_MESSAGES/messages.po
> msgmerge --update --previous locales/pt_BR/LC_MESSAGES/messages.po locales/messages.pot
> nox -s build -- -D language=pt_BR
> ```
> 
> It should work for any other language, replacing pt_BR with the proper language code.

Maybe you could add a new target for this to the noxfile? I guess some of the other translators could benefit from this.

**Comment by rffontenelle** [comment]
Strings containing hyperlinks *without a target* cannot be translated without Sphinx bringing up warning, which will always popup as error when *sphinx-build* has `-W` flag enabled (recommended when in CI).

e.g.: translating ``` `API token`_ ``` from guides/distributing-packages-using-setuptools.rst to ``` `token de API`_ ``` or ``` `token de API <token de API>`_ ``` will result in:

```
... /guides/distributing-packages-using-setuptools.rst:904: WARNING: inconsistent references in translated message. original: ['`API token`_'], translated: []
```

I'm thinking about translating all the guide in order to map all problematic cases for translation, opening a tracker issue and presenting PR with changes to work around them.

Any thoughts on the change?

**Comment by rffontenelle** [comment]
FWIW "Translations" GitHub Actions is failing to update POT with msg "protected branch hook declined", requiring an approval.

**Comment by di** [comment]
Here's a link to a failed run: https://github.com/pypa/packaging.python.org/runs/4290954151

**Comment by webknjaz** [comment]
> Strings containing hyperlinks _without a target_ cannot be translated without Sphinx bringing up warning, which will always popup as error when _sphinx-build_ has `-W` flag enabled (recommended when in CI).
> 
> e.g.: translating `` `API token`_ `` from guides/distributing-packages-using-setuptools.rst to `` `token de API`_ `` or `` `token de API <token de API>`_ `` will result in:
> 
> ```
> ... /guides/distributing-packages-using-setuptools.rst:904: WARNING: inconsistent references in translated message. original: ['`API token`_'], translated: []
> ```
> 
> _EDIT: Note, hyperlinks with a target would be `` `text <link>`_ ``_
> 
> I'm thinking about translating all the guide in order to map all problematic cases for translation, opening a tracker issue and presenting PR with changes to work around them.
> 
> Any thoughts on the change?

Have you tried keeping the original reference between angled brackets? As in:
```rst
`token de API <API token>`_
```

Tell me if it works and maybe I'd add more of those regex rules.

--------

> FWIW "Translations" GitHub Actions is failing to update POT with msg "protected branch hook declined", requiring an approval.

Yeah, I know, thanks. Pradyun and I have been brainstorming possible solutions on a few of our calls. GHA (and GH Apps) cannot skip branch protection (for various reasons). I wanted to apply a workaround but Pradyun didn't like the idea and suggested integrating `pre-commit.ci` so that it'd be autogenerated on each PR. But on our last meeting we realized that this may cause more problems. Pradyun suggested another solution — remove the POT file from the main branch and have an automation generating and committing it into a separate "translation" branch which would avoid conflicts. Apparently, Weblate is able to source the POT file from a different branch than it pushes the MO files back to.
We want to try it out but I haven't had time to actually experiment with this.

**Comment by webknjaz** [comment]
@meowmeowmeowcat FYI I've pinned this issue and turned it into a call to action.

**Comment by webknjaz** [comment]
@di could you go to https://readthedocs.org/dashboard/python-packaging-user-guide/translations/ and add `packagingpythonorg-pt-br` there? This one has a lot of the content translated already so we could start integrating it, I think.

**Comment by rffontenelle** [comment]
@webknjaz @di would you mind holding on? I'd rather proofread it first, if you don't mind. For now, I'm mostly fixing Sphinx warnings, not proofreading it.

**Comment by webknjaz** [comment]
As you wish, I just thought it'd be harmless to have it.

**Comment by rffontenelle** [comment]
@webknjaz oh well, feel free to enable pt-br. I guess in worst cases I will receive some feedback from other Brazilians 😄

**Comment by webknjaz** [comment]
FWIW we could just be silent about this and not announce until you feel it's ready :)

**Comment by rffontenelle** [comment]
Sure, make it a ln Easter egg :)

**Comment by rffontenelle** [comment]
Just to mention the Earth emoji was cut off in the issues view (at least on mobile device). 😕

Here's a screenshot:

![Screenshot_20211123-201658_Firefox](https://user-images.githubusercontent.com/1571783/143143825-ce07c38e-b4ef-44d4-ba70-619aeedd6ff7.jpg)

**Comment by rffontenelle** [comment]
> 
> Have you tried keeping the original reference between angled brackets? As in:
> 
> ```
> `token de API <API token>`_
> ```
> 
> Tell me if it works and maybe I'd add more of those regex rules.
> 

I'm using your suggestion already. For instance:

```
#: ../source/guides/distributing-packages-using-setuptools.rst:903
msgid ""
"Now you'll create a PyPI `API token`_ so you will be able to securely upload "
"your project."
msgstr ""
"Agora você criará um PyPI `token de API <API token>`_ para que possa enviar "
"o seu projeto com segurança."
```

and build output:

```
... /source/guides/distributing-packages-using-setuptools.rst:903: WARNING: inconsistent references in translated message. original: ['`API token`_'], translated: []
```

I don't think it is about the regex you set, as the resulting PO file received the correct translated content. It is Sphinx that is not accepting it as valid

**Comment by webknjaz** [comment]
> I'm using your suggestion already

I wasn't sure because in the previous comment you used this example:
```rst
`token de API <token de API>`_
```

Should we ask about this on the Sphinx's tracker?

**Comment by rffontenelle** [comment]
Indeed, the intention was to write like you did. Thanks for the clarification.

**Comment by webknjaz** [comment]
This should help us fix that failing workflow problem and autoupdate the source strings in Weblate: https://github.com/pypa/packaging.python.org/pull/1014.

**Comment by rffontenelle** [comment]
So, I'm trying to proofread pt-br translation, but POT in Weblate is not up to date. This causes the generated pages to have some untranslated strings. Please let me know when the POT in there is updated.

**Comment by webknjaz** [comment]
> @di could you go to [readthedocs.org/dashboard/python-packaging-user-guide/translations](https://readthedocs.org/dashboard/python-packaging-user-guide/translations/) and add `packagingpythonorg-pt-br` there? This one has a lot of the content translated already so we could start integrating it, I think.

@di need help here :point_up:

**Comment by webknjaz** [comment]
> POT in Weblate is not up to date. This causes the generated pages to have some untranslated strings. Please let me know when the POT in there is updated.

Apparently, I had a tiny bug in #1014 (wrong var name used due to copy-paste, I think). I've sent a follow-up hotfix #1021. Once that is approved and merged, Weblate should be able to pick up the updated POT file.

**Comment by rffontenelle** [comment]
@webknjaz something is not correct. I translated 67 strings from the 230 untranslated ones, and a new POT update, they are all blank again.

One symptom of this is that Weblate presents the check saying it was translated and presents my previous translation.

**Comment by webknjaz** [comment]
I may have messed it up a bit, there was a missynchronization due to removal of the translations from main and switching to the new branch. I'll try to see if I can recover that somehow.

**Comment by rffontenelle** [comment]
@webknjaz Nah, don't worry as I can simply reuse the previous translation. Actually I was more worried that the translation workflow was somehow causing the PO files to be replaced.

**Comment by webknjaz** [comment]
I've created the removal commit in https://github.com/pypa/packaging.python.org/pull/1023 (yet to be merged) and merged that commit to the translation branch + created a "revert" so that when the PR gets merged and the automation kicks in, Git would know to keep the files. And I also merged the Weblate repo branch (they have their own Git hosting for this). So somewhere in the middle of this process, I messed up (I think that was due to resolving a conflict incorrectly).

**Comment by di** [comment]
> @di could you go to readthedocs.org/dashboard/python-packaging-user-guide/translations and add `packagingpythonorg-pt-br` there? This one has a lot of the content translated already so we could start integrating it, I think.

Done!

**Comment by webknjaz** [comment]
Thanks! The only problem is that it doesn't seem to work now. The language selector appeared in the RTD's built-in version switcher but clicking it shows a 404. And using the canonical domain now redirects there too.

**Comment by webknjaz** [comment]
Looks like because the English site has "single version", it cannot properly route the translations: https://docs.readthedocs.io/en/stable/single_version.html. Without this, their webserver configuration seems to treat any URL prefix as a part of the main documentation project, which is English, and show the "Not Found" page.

@di @ewdurbin @pradyunsg I know that this has been enabled at some point in the past but it sounds like we have a valid reason to disable this for the sake of having multiple languages. This way, the URLs would be https://packaging.python.org/en/latest/ and https://packaging.python.org/pt_BR/latest/.

**Comment by ewdurbin** [comment]
Just checked and this setting is not enabled for the packaging.python.org project on readthedocs.org.

**Comment by webknjaz** [comment]
> Just checked and this setting is not enabled for the packaging.python.org project on readthedocs.org.

Are you sure? Is it behind the PSF's front-end, then?

**Comment by webknjaz** [comment]
Doesn't look like it is:
```console
$ dig packaging.python.org

; <<>> DiG 9.16.15 <<>> packaging.python.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3643
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;packaging.python.org.		IN	A

;; ANSWER SECTION:
packaging.python.org.	60	IN	CNAME	readthedocs.io.
readthedocs.io.		300	IN	A	104.17.33.82
readthedocs.io.		300	IN	A	104.17.32.82

;; Query time: 16 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Dec 06 13:36:07 CET 2021
;; MSG SIZE  rcvd: 109
```

**Comment by ewdurbin** [comment]
Unsure if someone got to this before me, but it appears to be working as desired?

**Comment by webknjaz** [comment]
@ewdurbin yep, looks like it. Thanks for checking! It seems I've been hitting in-browser cache and `Ctrl+Shift+R` helped.

@rffontenelle your translation is live! :tada: Do we want to advertise this already? Or do we want to do something about the language switcher? The RTD one works but the one coming from the upstream theme doesn't seem to be enabled/configured, I haven't looked into it but it may be worth checking.

**Comment by rffontenelle** [comment]
@webknjaz  Well, I don't know if the theme has any kind of support to switching languages, but I'm used to look for other languages in the RTD switcher. I'm fine as it is, but feel free to adjust the theme before spreading the word.

**Comment by webknjaz** [comment]
@rffontenelle the theme is derivative of the one used on https://docs.python.org and since it is now multilingual, I'd assume that our site could be as well. But it is indeed not that important *right now*. I think we could announce now and adjust later.

**Comment by rffontenelle** [comment]
@webknjaz I see. I don't know exactly how Python Docs switcher implementation works, but just to mention the switcher.js is created dynamically by [docsbuild-scripts' build_docs.py](https://github.com/python/docsbuild-scripts/blob/main/build_docs.py), which is run on daily-basis. This applies to the Python version and the languages available of the languages.

Maybe a chat with Julien Palard as he is the most active on this implementation, according to the commit history.

**Comment by webknjaz** [comment]
@rffontenelle I guess we can sort it over time. I don't care about this specific thing too much.

@di @ewdurbin there's a problem with the new setup that was raised in https://github.com/pypa/packaging.python.org/issues/1026 — could any of you implement that prefix redirect setting? Only you have the access to do this.

**Comment by ewdurbin** [comment]
Added prefix redirect. Does that resolve the issue?

```
ee-imac:~ ee$ curl -s -I  https://packaging.python.org/contribute/ | grep location
location: https://packaging.python.org/en/latest/contribute/
```

**Comment by webknjaz** [comment]
Works for me, thanks!

**Comment by webknjaz** [comment]
Just tweeted about this: https://twitter.com/webknjaz/status/1468240848450166797.

**Comment by webknjaz** [comment]
Hey @motok, I see you're almost finished with your translation. Could you share your RTD handle?

**Comment by webknjaz** [comment]
@motok?

**Comment by webknjaz** [comment]
@di @ewdurbin I think it's time to add Japanese officially. It's got higher percentage than Brazilian Portuguese which is already published. Could you do this?

**Comment by ewdurbin** [comment]
@webknjaz what is your RTD name? I think it'd be great to add you as a maintainer on the packaging.python.org RTD project so you can self-service these.

**Comment by ewdurbin** [comment]
I've added the Japanese project as a translation for now.

**Comment by webknjaz** [comment]
Same as GH — @webknjaz.

**Comment by ewdurbin** [comment]
> Same as GH — @webknjaz. Thanks!

invite sent.

**Comment by webknjaz** [comment]
@rffontenelle over time, the translations become stale. PT-BR probably needs updates, FYI.

**Comment by rffontenelle** [comment]
I'll work on it. Thanks, @webknjaz

**Comment by rffontenelle** [comment]
I noticed a ```:pep:`440` ``` flag is set in Weblate and was thinking that using a [regex check](https://docs.weblate.org/en/latest/user/checks.html#check-regex) ```regex:':pep:`[\d]+`'``` (haven't tested!) would be more efficient to all PEPs. Could this be tested? If it works, I can look around for other potential regex checks.

**Comment by willingc** [comment]
Start translating in your language at https://hosted.weblate.org/projects/pypa/packaging-python-org/

See additional information in the [Contributing Guide section for Translation](https://packaging.python.org/en/latest/contribute/#translations).

Updated: Nov. 17, 2023

*This info is also in the first post as well.*

**Comment by rffontenelle** [comment]
Hi all, I created #1431 as a way for testing translations and bringing up reST syntax errors in order to improve their quality. Feedback is welcome.

**Comment by rffontenelle** [comment]
@webknjaz Could you please mark the source string "Summary" with the [read-only](https://docs.weblate.org/en/latest/admin/checks.html#customizing-behavior-using-flags) flag in Weblate? The text is the title of [the core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/#summary), so I assume it should not be translated, but its translation propagates to warehouse's Summary which is a normal word and should be translated. You can find it by searching for `location:../source/specifications/core-metadata.rst:173` in the component's search box.

**Comment by webknjaz** [comment]
Apparently, the searches URLs seem linkable? https://hosted.weblate.org/translate/pypa/packaging-python-org/en/?q=location%3A..%2Fsource%2Fspecifications%2Fcore-metadata.rst%3A173&sort_by=-priority%2Cposition&checksum=

I wonder if this word can appear as a title in some later documents. Should we try coming up with special markup for things that are keywords...?

So... I checked https://hosted.weblate.org/translate/pypa/packaging-python-org/en/?q=location%3A..%2Fsource%2Fspecifications%2Fcore-metadata.rst%3A173&sort_by=-priority%2Cposition&checksum=#nearby and apparently I've already marked it as read-only in the past.

There's an alert
```
 Multiple failing checks 
Following checks are failing:
Inconsistent: Portuguese (Brazil), Portuguese (Brazil), Japanese, Japanese, Ukrainian, Ukrainian
```
there but I don't understand why it's showing up (and why is every language repeated twice there)...

Is this because some of the languages ended up submitting a translation earlier, before the phrase became read-only?

**Comment by webknjaz** [comment]
https://hosted.weblate.org/translate/pypa/packaging-python-org/uk/?q=location%3A..%2Fsource%2Fspecifications%2Fcore-metadata.rst%3A173&sort_by=-priority%2Cposition&checksum=

shows two alerts:
```
 Inconsistent 1
This string has more than one translation in this project or is untranslated in some components.
```
and
```
 Reused translation 2
Other source string: “Wrap up”
```
which is also weird.

**Comment by webknjaz** [comment]
Oh wait.. It says that the `read-only` flag is inherited for English. Adding the flag explicitly made it `read-only` across all languages.

**Comment by rffontenelle** [comment]
Maybe these metadata fields can be written as ``` ``literal text`` ``` to avoid matching with normal text?

The languages were showing up twice because Summary was propagated between two components. These languages probably translated in only one (e.g. pt_BR was not translated in warehouse) so the inconsistence check is for that language in both components.

**Comment by webknjaz** [comment]
> Maybe these metadata fields can be written as `` ``literal text`` `` to avoid matching with normal text?

Yeah, that's what I've been considering but wanted to hear more opinions..

> The languages were showing up twice because Summary was propagated between two components. These languages probably translated in only one (e.g. pt_BR did not have Summary translated in warehouse) so the inconsistency check is for that language in both components.

When you talk about components, do you mean PyPUG and Warehouse?

**Comment by rffontenelle** [comment]
> When you talk about components, do you mean PyPUG and Warehouse?

Yes. In the context of Weblate, "pypa" is a project that contains two components "warehouse" and "packaging.python.org".

**Comment by vldmrdev** [comment]
Hi! Could someone please clarify why it is not possible to switch to other translations on the page? I only see three languages: en, jr, pt-br, but there are other almost completed translations on the Webplate.

**Comment by webknjaz** [comment]
Because each needs to be deployed separately and I've only been doing that for languages that have completed translation at some point.
